core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -131,55 +136,99 @@ public BlockInfo getInfo(long blockId) throws IOException {
-      BlockLocationPolicy policy =
-          Preconditions.checkNotNull(options.getOptions().getUfsReadLocationPolicy(),
-              PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
-      GetWorkerOptions getWorkerOptions = GetWorkerOptions.defaults().setBlockId(info.getBlockId())
-          .setBlockSize(info.getLength()).setBlockWorkerInfos(getEligibleWorkers());
-      dataSource = policy.getWorker(getWorkerOptions);
-    } else { // Data will be read from Alluxio, determine which worker and if it is local
+    locations = locations.stream()
+        .filter(x -> workers.contains(x.getWorkerAddress())).collect(toList());
+    if (!locations.isEmpty()) {
+      // Case 1: we still have at least one worker containing the block.
       // TODO(calvin): Get location via a policy
       List<TieredIdentity> tieredLocations =
           locations.stream().map(location -> location.getWorkerAddress().getTieredIdentity())
-              .collect(Collectors.toList());
+              .collect(toList());
       Collections.shuffle(tieredLocations);
       Optional<TieredIdentity> nearest = mTieredIdentity.nearest(tieredLocations);
       if (nearest.isPresent()) {
-        dataSource = info.getLocations().stream()
-            .map(BlockLocation::getWorkerAddress)
-            .filter(a -> a.getTieredIdentity().equals(nearest.get()))
-            .findFirst().get();
+        dataSource = locations.stream().map(BlockLocation::getWorkerAddress)
+            .filter(a -> a.getTieredIdentity().equals(nearest.get())).findFirst().get();
         if (mTieredIdentity.getTier(0).getTierName().equals(Constants.LOCALITY_NODE)
             && mTieredIdentity.topTiersMatch(nearest.get())) {
           dataSourceType = BlockInStreamSource.LOCAL;
         } else {
           dataSourceType = BlockInStreamSource.REMOTE;
         }
       }
+    } else {
+      // Case 2: no worker contains the block; we must read from the UFS.
+      dataSourceType = BlockInStreamSource.UFS;
+      BlockLocationPolicy policy =
+          Preconditions.checkNotNull(options.getOptions().getUfsReadLocationPolicy(),
+              PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
+      blockWorkerInfo = blockWorkerInfo.stream()
+          .filter(x -> workers.contains(x.getNetAddress())).collect(toList());
+      GetWorkerOptions getWorkerOptions = GetWorkerOptions.defaults().setBlockId(info.getBlockId())
+          .setBlockSize(info.getLength()).setBlockWorkerInfos(blockWorkerInfo);
+      dataSource = policy.getWorker(getWorkerOptions);
     }
+
     if (dataSource == null) {
       throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
     }
-
     return BlockInStream.create(mContext, info, dataSource, dataSourceType, options);
   }
 
+  private Set<WorkerNetAddress> handleFailedWorkers(Set<WorkerNetAddress> workers,
+                                                    Map<WorkerNetAddress, Long> failedWorkers) {
```

Comment:
nit: indentation

Reply:
Just found and turned off my Intellij setting to auto align parameters. Fixed.

Reply:
You inspired me to do the same and finally fix that setting :)